anndata/compat.py

```diff
@@ -53,8 +53,13 @@ def _from_fixed_length_strings(value):
         if isinstance(
             dt_type, tuple
         ):  # vlen strings, could probably match better
-            dt_list[1] = "O"
-            new_dtype.append(tuple(dt_list))
+            # Fixing issues with h5py v2.10.0, could be cleaner
+            if issubclass(np.dtype(dt_type[0]).type, np.string_):
+                dt_list[1] = 'U{}'.format(int(dt_type[0][2:]))
+                new_dtype.append(tuple(dt_list))
+            else:
+                dt_list[1] = "O"
+                new_dtype.append(tuple(dt_list))
```

Comment:
The logic is a bit convoluted here. Simplified I think it looks like this:

```py

def _from_fixed_length_strings(value):
    """Convert from fixed length strings to unicode.

    For backwards compatability with older h5ad and zarr files.
    """
    new_dtype = []
    for dt in value.dtype.descr:
        dt_list = list(dt)
        dt_type = dt[1]
        # could probably match better
        is_vlen_string = isinstance(dt_type, tuple)
        if is_vlen_string:
            dt_type = dt_type[0]
        # Fixing issues with h5py v2.10.0, could be cleaner
        if issubclass(np.dtype(dt_type).type, np.string_):
            dt_list[1] = f"U{int(dt_type[2:])}"
        elif is_vlen_string:
            dt_list[1] = "O"
        new_dtype.append(tuple(dt_list))
    return value.astype(new_dtype)
```

right?

Reply:
Pretty much. It's a little unclear to me what exactly it's meant to mean when there's a tuple in the second field, but I think it's just generally annotation. Since we're pretty restrictive about what values we support in data frames, I think they tended to be fixed length strings. Just pushed a change cleaning this up.